src/directive.ts

```diff
@@ -0,0 +1,100 @@
+  DirectiveLocation,
+  GraphQLDirective,
+  GraphQLField,
+  GraphQLNonNull,
+  GraphQLResolveInfo,
+  GraphQLScalarType,
+  GraphQLSchema,
+  GraphQLString,
+} from 'graphql';
+import {
+  buildPathCache,
+  CasbinAuthTypeName,
+  CasbinContextEnforcerKey,
+  CasbinContextPathCacheKey,
+  CasbinContextRoleKey,
+  Info2Path,
+  PathAndAction2Key,
+} from './utils';
+import { EnforcerNotFoundError, PermissionInsufficientError } from './error';
+import { Enforcer } from 'casbin';
+
+export class CanDirective extends SchemaDirectiveVisitor {
+  static getDirectiveDeclaration(directiveName: string, schema: GraphQLSchema): GraphQLDirective | null | undefined {
+    return new GraphQLDirective({
+      name: 'can',
+      locations: [DirectiveLocation.FIELD_DEFINITION], // TODO: Add Support for DirectiveLocation.OBJECT
+      args: {
+        who: {
+          type: GraphQLString,
+          defaultValue: '*',
+        },
+      },
+    });
+  }
+
+  public visitFieldDefinition(field: GraphQLField<any, any>): GraphQLField<any, any> | void | null {
+    const { resolve = defaultFieldResolver } = field;
+    // if (!(field as any).hasCanDirective) {
+    //   (field as any).hasCanDirective = true;
+    // }
+    const { who } = this.args;
+    if (field.type instanceof GraphQLNonNull && field.type.ofType instanceof GraphQLScalarType) {
+      field.type = new GraphQLNonNull(new CanVisitType(field.type.ofType, who));
+    } else if (field.type instanceof GraphQLScalarType) {
+      field.type = new CanVisitType(field.type, who);
+    } else {
+      throw new Error(`Not a scalar type: ${field.type}`);
+    }
+
+    field.resolve = async function (source: any, args: { [p: string]: any }, context: any, info: GraphQLResolveInfo) {
+      const action = info.operation.operation;
+      const path = Info2Path(info);
+      let role = context[CasbinContextRoleKey];
+
+      if (!(CasbinContextEnforcerKey in context)) {
+        throw new EnforcerNotFoundError(`You must provide Enforcer in the context with key ${CasbinContextEnforcerKey}`);
+      }
+      const CasbinInstance = context[CasbinContextEnforcerKey] as Enforcer;
+      if (!(CasbinContextPathCacheKey in context)) {
+        context[CasbinContextPathCacheKey] = await buildPathCache(CasbinInstance);
+      }
+      const pathCache = context[CasbinContextPathCacheKey] as Array<string>;
+      if (!pathCache.includes(path)) {
+        await CasbinInstance.addPolicy(who, path, action);
+        pathCache.push(PathAndAction2Key(path, action));
+      }
+      const passed = await CasbinInstance.enforce(role, path, action);
+      if (passed) {
+        return await resolve.apply(this, [source, args, context, info]);
+      } else {
+        if (role === '*') {
+          role = 'anonymous';
+        }
+        throw new PermissionInsufficientError(`${role} can not ${action} ${path}`);
+      }
+    };
+  }
+
+  // public visitObject(object: GraphQLObjectType): GraphQLObjectType | void | null {}
```

Comment:
Remove commented lines before merge or add TODO, if something is missing.

Reply:
Yes, this is a pending work. I will add TODO